text/1008-tls-and-alpn.rst

```diff
@@ -0,0 +1,580 @@
+
+
+Motivation
+==========
+
+The Transport Layer Security (TLS) Protocol is widely used for providing
+privacy and data integrity between two communicating applications [1]_.
+EdgeDB as a database should naturally support TLS as the transport layer
+protocol.
+
+With the help of EdgeDB tooling, it will be easy to setup TLS
+certificates for both development and production environments. Given
+also that the benchmark [8]_ result of EdgeDB throughput overhead
+comparing TLS to vanilla TCP is about 5% ~ 15%, it is also proposed to
+communicate in TLS by default.
+
+.. image:: ./EdgeDB-TLS-Benchmark.png
+
+At last, with TLS enabled by default, it is possible to leverage the TLS
+Application-Layer Protocol Negotiation (ALPN) Extension [2]_ for secure
+and reliable protocol selection on top of the TLS transport, allowing
+EdgeDB to multiplex different frontend protocols like the binary
+protocol and the HTTP-based protocol on the same port as recommended in
+Pre-RFC 1910 [3]_.
+
+
+Overview
+========
+
+All connections to an EdgeDB server should use a transport with TLS
+enabled. An EdgeDB server instance needs at least a TLS certificate and
+its corresponding private key to start. A valid certificate and its key
+are usually expected to be provided to the server for production
+environments. EdgeDB clients (including the bindings and the CLI) must
+verify the authenticity and validity of the server certificate, or an
+error must be raised and the transport must be closed.
+
+To make the development experience streamlined, the EdgeDB tooling will
+automatically generate development-level certificates if none is
+provided. This process should be transparent and applies to both
+application and core developers.
+
+During the TLS handshake, the client may choose to negotiate a protocol
+using ALPN between the EdgeDB binary protocol and the HTTP protocol, and
+if succeeded that protocol must be used in the following communication.
+Otherwise (or if the client chose not to do ALPN), the HTTP protocol is
+supposed to be used as the default protocol.
+
+
+EdgeDB Server Certificate
+=========================
+
+When creating an EdgeDB instance (with either ``edgedb project init`` or
+``edgedb server init``), the user could provide a custom certificate and
+its private key to be used on the server::
+
+    --tls-certfile  Path to a single file in PEM format containing the
+                    certificate as well as any number of CA certificates
+                    needed to establish the certificate’s authenticity.
+                    If not present, a self-signed certificate will be
+                    generated and used instead (for development only!).
+    --tls-keyfile   Path to a file containing the private key. If not
+                    present, the private key will be taken from
+                    --tls-certfile as well.
+
+(This is exactly `ssl.SSLContext.load_cert_chain()` from Python [4]_.)
```

Comment:
```suggestion
(This is exactly ``ssl.SSLContext.load_cert_chain()`` from Python [4]_.)
```

Single backticks mean something else in ReST, please check if you use them elsewhere